# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/serviceworkers/NavigationPreloadManager.h
# Commit: afae3c50f7ac
# Full Hash: afae3c50f7acff60d9400adbdc3a8aa2f9d7bdfe
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2021-06-10 15:47:45
# Regressor Bug: 1564235
# File Overlap Count: 3
# Description:
#   Bug 1564235 - P2 Create NavigationPreloadManager.webidl and implement webidl binding. r=dom-worker-reviewers,webidl,asuth
#   
#   This patch creates NavigationPreloadManager.webidl and implements the webidl binding.
#   
#   Depends on D116913
# ==============================================================================

diff -r 2148ef30186a -r afae3c50f7ac dom/serviceworkers/NavigationPreloadManager.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/dom/serviceworkers/NavigationPreloadManager.h	Thu Jun 10 06:01:26 2021 +0000
@@ -0,0 +1,60 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=8 sts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef mozilla_dom_NavigationPreloadManager_h
+#define mozilla_dom_NavigationPreloadManager_h
+
+#include "nsCOMPtr.h"
+#include "nsCycleCollectionParticipant.h"
+#include "nsISupports.h"
+#include "nsWrapperCache.h"
+#include "mozilla/RefPtr.h"
+
+class nsIGlobalObject;
+
+namespace mozilla {
+namespace dom {
+
+class Promise;
+
+class NavigationPreloadManager final : public nsISupports,
+                                       public nsWrapperCache {
+ public:
+  NS_DECL_CYCLE_COLLECTING_ISUPPORTS
+  NS_DECL_CYCLE_COLLECTION_SCRIPT_HOLDER_CLASS(NavigationPreloadManager)
+
+  static bool IsValidHeader(const nsACString& aHeader);
+
+  explicit NavigationPreloadManager(nsCOMPtr<nsIGlobalObject>&& aGlobal);
+
+  // Webidl binding
+  nsIGlobalObject* GetParentObject() const { return mGlobal; }
+
+  JSObject* WrapObject(JSContext* aCx,
+                       JS::Handle<JSObject*> aGivenProto) override;
+
+  // WebIdl implementation
+  already_AddRefed<Promise> Enable();
+
+  already_AddRefed<Promise> Disable();
+
+  already_AddRefed<Promise> SetHeaderValue(const nsACString& aHeader);
+
+  already_AddRefed<Promise> GetState();
+
+ private:
+  ~NavigationPreloadManager() = default;
+
+  // General method for Enable()/Disable()
+  already_AddRefed<Promise> SetEnabled(bool aEnabled);
+
+  nsCOMPtr<nsIGlobalObject> mGlobal;
+};
+
+}  // namespace dom
+}  // namespace mozilla
+
+#endif  // mozilla_dom_NavigationPreloadManager_h